akka-stream/src/main/scala/akka/stream/impl/Sinks.scala

```diff
@@ -380,3 +359,142 @@ private[akka] final class SinkQueueAdapter[T](delegate: SinkQueue[T]) extends ak
   import akka.dispatch.ExecutionContexts.{ sameThreadExecutionContext ⇒ same }
   def pull(): CompletionStage[Optional[T]] = delegate.pull().map(_.asJava)(same).toJava
 }
+
+private[stream] object CancellablePublisherSink {
+  sealed trait PublisherSinkMessages[+T]
+  private case class RequestMore[T](subscriber: SubscriptionWithCursor[T], demand: Long) extends PublisherSinkMessages[T]
+  private case class Cancel[T](subscriber: SubscriptionWithCursor[T]) extends PublisherSinkMessages[T]
+  private case class Subscribe[T](subscriber: Subscriber[_ >: T]) extends PublisherSinkMessages[T]
+  private case object CancelPublisher extends PublisherSinkMessages[Nothing]
+
+}
+
+/**
+ * INTERNAL API
+ */
+final private[stream] class CancellablePublisherSink[T](drainWhenNoSubscribers: Boolean) extends GraphStageWithMaterializedValue[SinkShape[T], CancellablePublisher[T]] {
+
+  val in = Inlet[T]("CancellablePublisherSink.in")
+  override def initialAttributes = DefaultAttributes.cancellablePublisherSink
+  override val shape: SinkShape[T] = SinkShape.of(in)
+
+  override def toString: String = "CancellablePublisherSink"
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+    val InputBuffer(initBuffer, maxBuffer) = inheritedAttributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16))
+    require(maxBuffer > 0, "Max buffer size must be greater than 0")
+    require(initBuffer > 0, "Init buffer size must be greater than 0")
+
+    val stageLogic = new GraphStageLogic(shape) with CallbackWrapper[PublisherSinkMessages[T]] with SubscriberManagement[T] with InHandler {
+      private var downstreamDemand: Long = 0L
+      private var upstreamCompleted = false
+      private var failedWithError: Throwable = null
+
+      override type S = SubscriptionWithCursor[T]
+
+      override def maxBufferSize: Int = maxBuffer
+      override def initialBufferSize: Int = initBuffer
+
+      setHandler(in, this)
+
+      override def preStart(): Unit = {
+        setKeepGoing(true)
+        initCallback(callback.invoke)
+      }
+
+      override def postStop(): Unit = {
+        stopCallback {
```

Comment:
it's from CallbackWrapper. Contains logic of what materialized instance responce when stream is finished